### PR TITLE
Modularize nerium into base and contrib

### PR DESCRIPTION
**Pull request by henryrizzi**
This is an initial attempt to partially modularize nerium into separate submodules which could ideally be broken into separate repos if need be.

This PR should not be merged in it's current state, rather, it should be used to have a conversation about design decisions and improvements.

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -16,17 +16,18 @@ RUN set -ex && pip3 install pipenv --upgrade
...
 COPY --from=gcsfuse-build  /go/bin/gcsfuse /usr/local/bin/
 RUN pipenv install --system
 
-# $db_driver may be one of "postgres" or "mysql"
-ARG db_driver
 
-RUN if [ "$db_driver" = "postgres" ]; then apk add --no-cache py3-psycopg2; fi
-RUN if [ "$db_driver" = "mysql" ]; then pipenv install --system pymysql; fi
+RUN apk add --no-cache py3-psycopg2
+RUN pipenv install --system pymysql
```
Comment (tym-xqo) [inline Dockerfile]: If we're going to always load both of these, we should consolidate with the main install steps earlier.

File: README.md [hunk 2]
```diff
@@ -34,16 +34,16 @@ $ curl http://localhost:8081/<report_name>?<params>
...
-`/v1/<string:query_type>/<string:report_name>/?<params>`  
+`/v1/<string:query_extension>/<string:report_name>/?<params>`  
 `/v1/sql/<string:report_name>/<string:format>/?<params>`  
 `/v1/sql/<string:report_name>/compact/?<params>`
 
 `report_name` should match the name of a given SQL file, minus the `.sql` extension. Params are as specified in the queries themselves.
 
-`query_type` is an optional file extension string and defaults to 'sql'
-`format` is an optional formatter name, and defaults to 'default'. 'compact' is also available, as noted above. (Note that while optional, if you wish to specify a format, you must include the query_type also.)
+`query_extension` is an optional file extension string and defaults to 'sql'
```
Comment (tym-xqo) [inline README.md:43]: Now that we have `bankend_lookup` method, we can and should remove query_extension from the URL pattern.

File: app.py [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+            # TODO: add logging and log a warning here
+            #     and/or return format not found message to client(?)
+            payload = query_result
+        return payload
+
+
+api.add_resource(
+    ReportAPI,
+    '/v1/<string:report_name>/',
+    '/v1/<string:query_type>/<string:report_name>/',
```
Comment (ghost) [inline app.py]: should this be `query_extension`?

File: README.md [hunk 4]
```diff
@@ -34,16 +34,16 @@ $ curl http://localhost:8081/<report_name>?<params>
...
+`format` is an optional formatter name, and defaults to 'default'. 'compact' is also available, as noted above. (Note that while optional, if you wish to specify a format, you must include the query_extension also.)
```
Comment (tym-xqo) [inline README.md:44]: `format` might work better as a query string parameter

File: app.py [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+    def get(self, report_name, query_type='sql', format_='default'):
+        # Lookup query_type and fetch result from corresponding class
+        try:
+            result_cls = self.query_type_lookup[query_type]
+        except KeyError:
+            result_cls = SQLResultSet
+
+        loader = result_cls(report_name,
+                            **request.args.to_dict())
+        query_result = loader.result()
```
Comment (tym-xqo) [inline app.py]: @henryrizzi We need to change this to call the `result_set` method instead. I can take that.

**Review by ghost — COMMENTED** [review]
lgtm

**Review by tym-xqo — CHANGES_REQUESTED** [review]

**Review by tym-xqo — APPROVED** [review]

**Review by luzmcosta — APPROVED** [review]
👍 I'm trying to get more familiar with this repo, so thanks for calling this out to me!

File: Makefile [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+	docker push us.gcr.io/lexical-cider-93918/nerium:latest
+	kubectl create secret generic nerium-dotenv --from-file=$(SECRET_PATH)nerium/.env
+	kubectl create -f nerium-k8s.yml
+	kubectl set image deployments/nerium nerium=us.gcr.io/lexical-cider-93918/nerium:latest
+
+	kubectl expose deployment nerium --name=results --type=NodePort --port=80 --target-port=8081
+
+destroy-results:
+	gcloud container clusters get-credentials playground
+	kubectl delete deployment nerium
```
Comment (luzmcosta) [inline Makefile:15]: I love a good teardown section. 😄

File: app.py [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+        if isinstance(obj, decimal.Decimal):
+            return str(obj)
+        elif hasattr(obj, 'isoformat'):
+            return obj.isoformat()
+        else:
+            return obj
+        return JSONEncoder.default(self, obj)
+
+
+app.config['RESTFUL_JSON'] = {'cls': ResultJSONEncoder}
```
Comment (luzmcosta) [inline app.py:30]: Curious: what does "cls" stand for here?

Comment (tym-xqo) [reply app.py:30]: @luzmcosta Class. This is essentially replacing the default Python `json` encoder (which returns 'not serializable' errors for decimals and datetimes) with our custom one.

**Comment by tym-xqo** [comment]
@henryrizzi I know this was opened as WIP, but I suggest we merge at this point. It's not perfect, but anything I would change feels like work for a fresh iteration, to me. What do you think?

File: app.py (hunk 7 above)
Comment (luzmcosta) [reply app.py:30]: 👍 🙇

File: nerium/contrib/resultset/takei.py [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+        return sql_query
+
+    def result(self):
+        backend_path = self.get_query_path().parent
+        backend_template = self.backend_lookup(backend_path)
+        takei_pwd = os.getenv('MYSQL_PWD', None)
+        backend = backend_template.format(password=takei_pwd)
+        db = records.Database(backend)
+
+        client_id = self.kwargs.get('client_id', 0)
```
Comment (henryrizzi) [inline nerium/contrib/resultset/takei.py]: Does this seem like a reasonable thing to do when `client_id` is not supplied as a kwarg? I could change this to some other value, but it at least allows table_substitution to run. @tym-oao

**Pull request merged**